INT-2166: Add SecurityContext Propagation

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-2166
- Introduce `ThreadStatePropagationChannelInterceptor` based on the `ExecutorChannelInterceptor`
- Add `SecurityContextPropagationChannelInterceptor`,`SecurityContextCleanupChannelInterceptor`
- Introduce `AbstractExecutorChannel` to utilize `ExecutorChannelInterceptor` logic
- Introduce `MessageHandlingTaskDecorator` to avoid package tangle from `dispatcher` and `channel`
- Introduce `SecurityContextCleanupAdvice` for those cases when we don't get deal with `MessageChannel`s already, but want to have proper way to cleanup `SecurityContext`
- Make `GlobalChannelInterceptorProcessor` as `SmartInitializingSingleton` to avoid `phase` conflicts.
- Fix `MessagingAnnotationPostProcessor` to use `beanFactory.initializeBean(endpoint, endpointBeanName);` instead of manual `start()` invocation bypassing the `phase` logic, hence having a bug, when endpoints have been started very early
- Optimise `AbstractPollableChannel` to use `size` field from `ChannelInterceptorList` instead of `size()` from `Collection<?>`
- Fix `AnnotatedEndpointActivationTests` extracting separate component for annotation configuration instead of using test class directly. This caused very late Messaging Annotations process on that class
- Fix typo in the `spring-integration-jdbc-4.2.xsd`
- Remove some `SOUT`s throughout the project

TODO Docs

artembilan committed:
INT-2166: Add SecurityContext Propagation

JIRA: https://jira.spring.io/browse/INT-2166

* Introduce `ThreadStatePropagationChannelInterceptor` based on the `ExecutorChannelInterceptor`
* Add `SecurityContextPropagationChannelInterceptor`,`SecurityContextCleanupChannelInterceptor`
* Introduce `AbstractExecutorChannel` to utilize `ExecutorChannelInterceptor` logic
* Introduce `MessageHandlingTaskDecorator` to avoid package tangle from `dispatcher` and `channel`
* Introduce `SecurityContextCleanupAdvice` for those cases when we don't get deal with `MessageChannel`s already, but want to have proper way to cleanup `SecurityContext`
* Make `GlobalChannelInterceptorProcessor` as `SmartInitializingSingleton` to avoid `phase` conflicts.
* Fix `MessagingAnnotationPostProcessor` to use `beanFactory.initializeBean(endpoint, endpointBeanName);` instead of manual `start()` invocation bypassing the `phase` logic, hence having a bug, when endpoints have been started very early
* Optimise `AbstractPollableChannel` to use `size` field from `ChannelInterceptorList` instead of `size()` from `Collection<?>`
* Fix `AnnotatedEndpointActivationTests` extracting separate component for annotation configuration instead of using test class directly. This caused very late Messaging Annotations process on that class
* Fix typo in the `spring-integration-jdbc-4.2.xsd`
* Remove some `SOUT`s throughout the project

TODO Docs

PR Comments:

* Remove redundant `AbstractExecutorChannel#executorInterceptors` and make logic based on the `super.interceptors`
 * Fix wrong imports order
 * JavaDocs for `ThreadStatePropagationChannelInterceptor`
 * Docs for `SecurityContext` propagation

INT-3593: Fix FTP PartialSuccess Tests

JIRA: https://jira.spring.io/browse/INT-3593

Sort the files for the MPUT tests.

INT-2166: Add SecurityContext Propagation

JIRA: https://jira.spring.io/browse/INT-2166

* Introduce `ThreadStatePropagationChannelInterceptor` based on the `ExecutorChannelInterceptor`
* Add `SecurityContextPropagationChannelInterceptor`,`SecurityContextCleanupChannelInterceptor`
* Introduce `AbstractExecutorChannel` to utilize `ExecutorChannelInterceptor` logic
* Introduce `MessageHandlingTaskDecorator` to avoid package tangle from `dispatcher` and `channel`
* Introduce `SecurityContextCleanupAdvice` for those cases when we don't get deal with `MessageChannel`s already, but want to have proper way to cleanup `SecurityContext`
* Make `GlobalChannelInterceptorProcessor` as `SmartInitializingSingleton` to avoid `phase` conflicts.
* Fix `MessagingAnnotationPostProcessor` to use `beanFactory.initializeBean(endpoint, endpointBeanName);` instead of manual `start()` invocation bypassing the `phase` logic, hence having a bug, when endpoints have been started very early
* Optimise `AbstractPollableChannel` to use `size` field from `ChannelInterceptorList` instead of `size()` from `Collection<?>`
* Fix `AnnotatedEndpointActivationTests` extracting separate component for annotation configuration instead of using test class directly. This caused very late Messaging Annotations process on that class
* Fix typo in the `spring-integration-jdbc-4.2.xsd`
* Remove some `SOUT`s throughout the project

TODO Docs

PR Comments:

* Remove redundant `AbstractExecutorChannel#executorInterceptors` and make logic based on the `super.interceptors`
 * Fix wrong imports order
 * JavaDocs for `ThreadStatePropagationChannelInterceptor`
 * Docs for `SecurityContext` propagation

Doc Polishing

artembilan commented:
@garyrussell , it is an unexpected big change, so I guess you won't want to review it and merge for tomorrow's release. 
Anyway let me know if you like it and I go ahead with Docs.
Thanks

garyrussell commented:
Yeah - I've pushed this to RC1 but I will start reviewing it; we can decide later if we can pull it back into M2.

Let me know if you don't think we can merge all my PRs today; at a minimum I would like to include zookeeper in M2 so people can start trying it.

garyrussell commented on spring-integration-security/src/main/java/org/springframework/integration/security/channel/SecurityContextPropagationChannelInterceptor.java:
```diff
@@ -0,0 +1,174 @@
+			}
+			this.messageBuilderFactorySet = true;
+		}
+		return this.messageBuilderFactory;
```
Can be null; `AssertState` ?

artembilan replied:
No, it can't be `null`. See `IntegrationUtils.getMessageBuilderFactory()` implementation.

garyrussell replied:
?? it will be null if there's no BF...

    if (this.beanFactory != null) {
   			this.messageBuilderFactory =

artembilan replied:
??? But we have this above:
```
private volatile MessageBuilderFactory messageBuilderFactory = new DefaultMessageBuilderFactory();
```
So, if there is no `BF` we don't override the default value

garyrussell replied:
Ah; missed that; sorry; it might be clearer is all the code was in the same place.

artembilan replied:
Yeah... We can change this to more clearer code, but it is so simple change that I don't want to worry about it.
More over the copy/paste from existing classes :smile:

garyrussell replied:
Sure, np; I need to study this whole PR some more so will have to wait for RC1.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/ExecutorChannel.java:
```diff
@@ -124,15 +106,28 @@ public final void onInit() {
+
+			@Override
+			public Runnable decorate(MessageHandlingRunnable task) {
+				return new MessageHandlingTask(task);
```
I need to think about this some more but, given all the streamlining we've been doing for XD lately, I am not keen on all this per-message extra object creation, especially when there are no interceptors.

artembilan replied:
Well, I think `if...else` for the interceptors presence must convince you :smile:

artembilan commented:
Pushed after rebase

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/ExecutorChannel.java:
```diff
@@ -124,15 +106,33 @@ public final void onInit() {
+
+			@Override
+			public Runnable decorate(MessageHandlingRunnable task) {
+				if (!ExecutorChannel.this.executorInterceptors.isEmpty()) {
```
See the optimization I did for INT-3719; calling `size()` was much more expensive than looking up a field. However, that might have been because it was a `CopyOnWriteArrayList`.

However, yours is a simple `ArrayList` - should it be a COWAL?

Not saying it's wrong but we need to be very sensitive about putting any unnecessary code in the main message flow.

artembilan replied:
OK. Sounds reasonable.
If the core interceptors logic is done around `CopyOnWriteArrayList` and `size`, this one for `ExecutorChannelInterceptor`s must be similar.

Will rework on RC1 dead line!

garyrussell commented:
I don't really like that the security context propagation is tied to a secured channel; it seems to me that thread-state propagation (of any type) is orthogonal to whether the channel itself is secured.

Also, we are leaking the header used to propagate the context into user space.

I was expecting a much simpler interceptor, something like...

````java
public class SecPropInterceptor extends ChannelInterceptorAdapter {

	@SuppressWarnings("unchecked")
	@Override
	public Message<?> preSend(Message<?> message, MessageChannel channel) {
		return new MessageWithSecurityContext((Message<Object>) message,
				SecurityContextHolder.getContext().getAuthentication());
	}

	@Override
	public Message<?> postReceive(Message<?> message, MessageChannel channel) {
		MessageWithSecurityContext messageWithSecurityContext = (MessageWithSecurityContext) message;
		SecurityContextHolder.getContext().setAuthentication(messageWithSecurityContext.state);
		return messageWithSecurityContext.message;
	}

	public Message<?> preHandle(Message<?> message, MessageChannel channel) {
		return postReceive(message, channel);
	}

	private static class MessageWithThreadState<S> implements Message<Object> {

		final Message<Object> message;

		final S state;


		public MessageWithThreadState(Message<Object> message, S state) {
			this.message = message;
			this.state = state;
		}

		@Override
		public Object getPayload() {
			throw new UnsupportedOperationException();
		}

		@Override
		public MessageHeaders getHeaders() {
			throw new UnsupportedOperationException();
		}

	}

	private static class MessageWithSecurityContext extends MessageWithThreadState<Authentication> {

		public MessageWithSecurityContext(Message<Object> message, Authentication state) {
			super(message, state);
		}

	}

}
````

A strategy like this could allow any thread-bound context to be transferred via the channel. Of course, it would have to be the last interceptor in the chain.

It's also not clear to me how your security stack stuff works on the `receive()` side - how does the `deque` ever get cleaned up there?

artembilan commented:
Pushed

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractExecutorChannel.java:
```diff
@@ -0,0 +1,221 @@
+
+	@Override
+	public void addInterceptor(int index, ChannelInterceptor interceptor) {
+		super.addInterceptor(index, interceptor);
```
This is a little peculiar - does the index represent the position in super interceptors or executor interceptors?

Maybe we should only support 0 as the index here?

garyrussell replied:
(if it's an exec interceptor)

artembilan replied:
Why don't you like my `if...else` logic around `index` below?
We can go ahead with your `0` advice and document it in the JavaDocs.
From other side I think I can figure out how to do the same only with `super.interceptors` without this extra `executorInterceptors`.
WDYT?

garyrussell replied:
The problem is that someone has 4 interceptors (normal), 2 exec and they want to insert a new interceptor in the middle.

Do they use "2" or "1" ? 

It's impossible.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractExecutorChannel.java:
```diff
@@ -0,0 +1,221 @@
+			return true;
+		}
+		else {
+			return false;
```
Should be true if not an exec. inter. and super.remove() was successful

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/interceptor/ThreadStatePropagationChannelInterceptor.java:
```diff
@@ -0,0 +1,119 @@
+	public final Message<?> preSend(Message<?> message, MessageChannel channel) {
+		S threadContext = obtainPropagatingContext(message, channel);
+		if (threadContext != null) {
+			return new MessageWithThreadState<S>(message, threadContext);
```
We will have to be careful to document that only one of these interceptors are allowed and it must be the last interceptor in the list. Otherwise a subsequent interceptor might remove the wrapper.

artembilan replied:
Well. Maybe we should introduce something like `CompositeThreadStatePropagationChannelInterceptor` to allow to inject several similar interceptors anyway.
Of course, end-user always can implement his own logic in his extension of this class, but what he will be able to do with those out-of-the-box implementations... 
If you'd like, I'll try to figure out how to be.

artembilan replied:
@garyrussell , actually we are OK with the current solution.
Consider we have two of such interceptors. The first performs its `preSend` and wraps `Message<?>` to the `MessageWithThreadState` which is `Message<?>`, too!
The next one in the list does the same and wraps our `MessageWithThreadState` to another one.

Now we are reaching `postReceive` (`beforeHandle`). Having the stack there we perform the second interceptor for the `MessageWithThreadState` with `message` as the `MessageWithThreadState` and unwrap it before return.
The first interceptor in the stack unwraps `MessageWithThreadState` to the original `Message<?>` normally.

Have I missed anything?

garyrussell replied:
Yes, but what if the next interceptor is not one of these; and does MessageBuilder.withPayload(...)...build()  ???

artembilan replied:
Well that is [Clear and present danger](https://en.wikipedia.org/wiki/Clear_and_Present_Danger_%28film%29) for any interceptor which modifies message.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/config/annotation/AbstractMethodAnnotationPostProcessor.java:
```diff
@@ -72,6 +70,8 @@
 
+import org.aopalliance.aop.Advice;
+
 /**
```
Do you have the proper settings on your Mac IDEA?

artembilan replied:
What do you mean here?
I have imports settings like `org.*` is coming after `org.springframework.*`
Or don't you like here that I have auto-optimization for imports?

garyrussell replied:
Maybe I need to review my settings - my IDE moves it back - I need to check against the SF wiki doc.

garyrussell commented:
Otherwise LGTM; need to review tests, but you can go ahead with docs.

artembilan commented:
Pushed

garyrussell commented:
Doc polishing here: https://github.com/garyrussell/spring-integration/commit/2f4dcfa8cc70e69e75125589ff8021345a96a735

garyrussell commented on spring-integration-security/src/main/java/org/springframework/integration/security/context/SecurityContextCleanupChannelInterceptor.java:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	@Override
+	public void afterSendCompletion(Message<?> message, MessageChannel channel, boolean sent, Exception ex) {
```
Why are we cleaning up on the sending side?

artembilan replied:
Consider the case:
````xml
<channel id="input">
    <queue/>
    <interceptors> 
	<beans:bean
					class="org.springframework.integration.security.context.SecurityContextPropagationChannelInterceptor"/>
    </interceptors>
</channel>
...
<channel id="output">
    <queue/>
</channel>
````

So, with `preSend()` on the `input` we propagate `SecurityContext` to that Scheduler's Thread.
And on the `send()` of the `output` channel we leave that Thread. 

So, nothing helps us here to free that thread from the current ctx for the future.

You have noticed good case with your second comment: we really can mark `<poller>` with `SecurityContextCleanupAdvice`. And we can address that in the documentation and its JavaDocs. But looks like nothing helps us with the `ExecutorChannel` on it `send()` part...

garyrussell replied:
This doesn't make sense to me (a propagating interceptor on a direct channel). It should only be allowed when an executor or queue is involved.

artembilan replied:
Sorry, Gary, I'm not sure here in your comment.
You suggest restrict `SecurityContextPropagationChannelInterceptor` for `DirectChannel`s or even `SecurityContextCleanupChannelInterceptor`? 
The first one just adds a bit more overhead for the message flow performance wrapping message.
But the second - `SecurityContextCleanupChannelInterceptor` - ... is more critical and may be dangerous because we continue with the same Thread but already without security ctx.
Let me know you position here.
Thanks

artembilan replied:
Although `SecurityContextPropagationChannelInterceptor` has that restriction:
````java
@Override
protected Authentication obtainPropagatingContext(Message<?> message, MessageChannel channel) {
	if (!DirectChannel.class.isAssignableFrom(AopUtils.getTargetClass(channel))) {
		return SecurityContextHolder.getContext().getAuthentication();
	}
	return null;
}
````
So, I guess you mean here just only `SecurityContextCleanupChannelInterceptor`.
Addressing now...
If that will continue discussion after my next push.
Pulling your Docs polishing BTW.

garyrussell replied:
It just makes no sense to me to "propagate" the security context on a channel where no async handoff to another thread occurs.

artembilan replied:
??? So, I really don't understand what you worry here about.
The `SecurityContextPropagationChannelInterceptor` doesn't do that because of the `if` in its `obtainPropagatingContext()`.
The `SecurityContextCleanupChannelInterceptor` doesn't propagate anything. But I agree that it is wrong to allow to clean up context on the `DirectChannel`, so I'm fixing that now.
What else have I missed from your perspective?

garyrussell replied:
Right, it doesn't do it, but it "looks weird" to have such an interceptor declared on a `DirectChannel` as in your 'consider this case' above. It should be on the queue channel.

garyrussell commented on src/reference/asciidoc/security.adoc:
```diff
@@ -93,3 +102,108 @@ public class ContextConfiguration {
+
+<int:outbound-channel-adapter channel="queueChannel" ref="securedService">
+    <int:request-handler-advice-chain>
+    	<bean class="org.springframework.integration.security.context.SecurityContextCleanupAdvice"/>
```
This will clean up immediately. If there's an `output-channel` and they want to SECURE the whole flow, the advice has to go on the poller instead of in the RHAC.

artembilan replied:
That's correct. And it makes this `Advice` so flexible :smile: 
For those cases like `output-channel` we have `SecurityContextCleanupChannelInterceptor` with its cleanup features on those places when we leave current thread throughout the channel: `send()` in most cases (`queue` and `executor`).

artembilan commented:
The docs polishing looks good to me, BTW.
Thanks

artembilan commented:
> I think I'd prefer a single interceptor for this use case.

Agreed. I can add boolean option to the `SecurityContextPropagationChannelInterceptor` to utilize clean up function as well.

artembilan committed:
Address PR comments

artembilan commented:
Pushed

artembilan commented:
I was forced to rebase and squesh your commit because our branches have had a different history.
So, merged and rebased and pushed after conflict resolving

garyrussell commented on spring-integration-ftp/src/test/java/org/springframework/integration/ftp/outbound/FtpServerOutboundTests-context.xml:
```diff
@@ -155,4 +157,7 @@
 
+	<bean id="sortingFilter"
+		class="org.springframework.integration.ftp.outbound.FtpServerOutboundTests$SortingFileListFilter"/>
+
```
Why is this here? Rebase needed?

garyrussell commented on spring-integration-security/src/main/java/org/springframework/integration/security/context/SecurityContextCleanupChannelInterceptor.java:
```diff
@@ -0,0 +1,64 @@
+		if (!DirectChannel.class.isAssignableFrom(AopUtils.getTargetClass(channel))) {
+			SecurityContextCleanupAdvice.cleanup();
+		}
+	}
```
I still fail to understand why we ever need cleanup on the sending side; please convince me.

We're handing the context to the next thread; I don't see why that ever affects the current thread.

artembilan replied:
Pushed after rebase.
I worry here about inter-flow threads when we have a `QueueChannel` as input to the that sub-flow, and `QueueChannel` as well as output. 
So, we must propagate context from this sub-flow to the other side of the 'output' and ... clean up it here in this thread!
From other side I'd say that only `afterSendCompletion` is a case when we should allow to clean up context by the `ChannelInterceptor`. So, end-user knows that the context is cleaned exactly on after the invocation within this thread, which is really initiated by the non-DirectChannel `send()`.

garyrussell replied:
The problem is that cleaning up at that point is incorrect - what about a recipient list router where the first subscriber sends to a queue channel? We'll lose the security context for subsequent subscribers.

I think it's easier to understand if the poller does the cleanup when the thread returns to the TE.

artembilan replied:
OK. That's good argument.
So, let me get rid of this `SecurityContextCleanupChannelInterceptor` since it does not make so much sense, unlike more general `SecurityContextCleanupAdvice`, which can be used for such a custom `ChannelInterceptor`, BTW...
Let me know if it OK with I got your point here.

Thank you for such a wisdom feedback!

garyrussell replied:
:+1:

artembilan committed:
Address PR comments

* Extract `ExecutorChannelInterceptor` logic in the `PollingConsumer`
to have an ability to invoke `afterMessageHandled()` on the TaskScheduler's Thread
for example for the `SecurityContext` clean up
* Get rid of all that redundant "clean up" stuff
* Docs polishing

artembilan commented:
Pushed

artembilan commented:
I see I missed to test project locally :cry:

garyrussell commented:
I am seeing the same failures as travis

artembilan commented:
Fixed, rebased and pushed

artembilan committed:
Fix `NPE` in the `PollingConsumer`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/endpoint/PollingConsumer.java:
```diff
@@ -71,15 +91,88 @@ protected void doStop() {
+		if (this.channelInterceptors != null) {
+			for (ChannelInterceptor interceptor : this.channelInterceptors) {
+				if (interceptor instanceof ExecutorChannelInterceptor) {
+					executorInterceptorPresent = true;
```
I think we need an optimization to avoid iterating over the interceptors to see if we have an exec interceptor.

Maybe a sub interface of `CIA` - `ExecutorChannelInterceptorAware.hasExecInterceptor()` ?

We can still get the list up front, but instead of iterating, just check the boolean.

artembilan replied:
Sounds reasonable.
Will try and come back to you tomorrow (or in the evening) on the matter.
Thanks

artembilan commented:
Pushed

artembilan committed:
Introduce `ExecutorChannelInterceptorAware` to avoid iterators on each message

garyrussell commented on spring-integration-amqp/src/main/java/org/springframework/integration/amqp/channel/PollableAmqpChannel.java:
```diff
@@ -40,8 +43,8 @@
-public class PollableAmqpChannel extends AbstractAmqpChannel implements PollableChannel,
-		PollableChannelManagement {
+public class PollableAmqpChannel extends AbstractAmqpChannel
+		implements PollableChannel, PollableChannelManagement, ExecutorChannelInterceptorAware {
```
Unfortunately, this name now looks wrong here.

We can't change `ExecutorChannelInterceptor` (which is also technically wrong), but maybe we could call this `HandlerInvokerInterceptorAware` WDYT?

garyrussell replied:
That name's no good either (a DirectChannel is a handler invoker too) - I guess we can live with it as is.

artembilan replied:
`ExecutorTaskInterceptorAware `?
Because `PollableChannel` works withing the executor's task anyway

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/endpoint/PollingConsumer.java:
```diff
@@ -68,18 +87,82 @@ protected void doStop() {
+			String description = "Failed to handle " + message + " to " + this + " in " + this.handler;
+			throw new MessageDeliveryException(message, description, ex);
+		}
+		catch (Error ex) {
```
Add `//NOSONAR - ok because we re-throw it below`

Otherwise Sonar will bark at us.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractExecutorChannel.java:
```diff
@@ -0,0 +1,209 @@
+				String description = "Failed to handle " + message + " to " + this + " in " + messageHandler;
+				throw new MessageDeliveryException(message, description, ex);
+			}
+			catch (Error ex) {
```
//NOSONAR here too

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractExecutorChannel.java:
```diff
@@ -0,0 +1,209 @@
+					interceptor.afterMessageHandled(message, AbstractExecutorChannel.this,
+							this.delegate.getMessageHandler(), ex);
+				}
+				catch (Throwable ex2) {
```
//NOSONAR - intentional

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractExecutorChannel.java:
```diff
@@ -0,0 +1,209 @@
+				}
+				messageHandler.handleMessage(message);
+				if (!CollectionUtils.isEmpty(interceptorStack)) {
+					triggerAfterMessageHandled(message, null, interceptorStack);
```
Since we're eating exceptions in this method, perhaps it should return the last exception (after the while loop completes) so we can re-throw it here.

garyrussell replied:
Actually, I guess it's ok, since the thread is being returned to the executor at this point anyway (same thing for the poller).

garyrussell commented:
That's all; I need to do a little polish on the docs; then I'll merge Monday - I can fix the //NOSONAR issues too.

artembilan commented:
Thank you, Gary!
I will be appreciated :smile:

garyrussell commented:
Final? Polish here: https://github.com/garyrussell/spring-integration/commit/860f335296771a2a99b73393d134fcfdfbb2ece8

garyrussell commented:
Merged as 09fb4f78c9ffce753abd035f7754764d0a61ffa4
